### PR TITLE
Update documentation for Vec::into_boxed_slice to be more clear about excess capacity

### PR DESCRIPTION
**Pull request by invpt**
Currently, the documentation for Vec::into_boxed_slice says that "if the vector has excess capacity, its items will be moved into a newly-allocated buffer with exactly the right capacity." This is misleading, as copies do not necessarily occur, depending on if the allocator supports in-place shrinking. I copied some of the wording from shrink_to_fit, though it could potentially still be worded better than this.

### CODE REVIEW

File: library/alloc/src/vec/mod.rs [hunk 1]
```diff
@@ -1074,8 +1074,9 @@ impl<T, A: Allocator> Vec<T, A> {
 
     /// Converts the vector into [`Box<[T]>`][owned slice].
     ///
-    /// If the vector has excess capacity, its items will be moved into a
-    /// newly-allocated buffer with exactly the right capacity.
+    /// If the vector has excess capacity, it will shrink down as close as
+    /// possible to the length but the allocator may still inform the vector
+    /// that there is space for a few more elements.
```
Comment (the8472) [inline library/alloc/src/vec/mod.rs:1079]: I don't think the "few more elements" part is relevant because it'll still [fit](https://doc.rust-lang.org/nightly/std/alloc/trait.Allocator.html#memory-fitting) the Box layout.
I think it should just link to `Allocator::shrink` which can explain the details.

Comment (invpt) [reply library/alloc/src/vec/mod.rs:1079]: Yes, I was just looking at that bit about layout fitting. It might be good to avoid linking to low-level documentation for a higher-level method though (without a clear, high-level explanation). Another option is to just link to shrink_to_fit, and maybe to improve the shrink_to_fit documentation and have *that* link to Allocator::shrink.

Comment (invpt) [reply library/alloc/src/vec/mod.rs]: Made the changes according to what I said above.

File: library/alloc/src/vec/mod.rs [hunk 2]
```diff
@@ -1023,8 +1023,10 @@ impl<T, A: Allocator> Vec<T, A> {
 
     /// Shrinks the capacity of the vector as much as possible.
     ///
-    /// It will drop down as close as possible to the length but the allocator
-    /// may still inform the vector that there is space for a few more elements.
+    /// The behavior of this method depends on the allocator being used. If the allocator supports
+    /// in-place shrinking, some or all of the excess capacity may be returned to the allocator.
+    /// If not, this method may cause the vector's data to be copied to a new allocation with
+    /// exactly the correct size. See [`Allocator::shrink`] for more details.
```
Comment (the8472) [inline library/alloc/src/vec/mod.rs:1029]: We don't guarantee "exactly" though. The same fuzziness as `with_capacity` or `reserve_exact` applies.

File: library/alloc/src/vec/mod.rs [hunk 3]
```diff
@@ -1074,10 +1076,10 @@ impl<T, A: Allocator> Vec<T, A> {
 
     /// Converts the vector into [`Box<[T]>`][owned slice].
     ///
-    /// If the vector has excess capacity, its items will be moved into a
-    /// newly-allocated buffer with exactly the right capacity.
+    /// Before doing the conversion, this method calls [`shrink_to_fit`] to remove excess capacity.
```
Comment (the8472) [inline library/alloc/src/vec/mod.rs:1079]: Rather that documenting how it's implementing (calling a specific method) it should describe behavior. "discards excess capacity, like shrink_to_fit" or something like that.

File: library/alloc/src/vec/mod.rs (hunk 2 above)
Comment (invpt) [reply library/alloc/src/vec/mod.rs]: Got it, this claim was made in at least one other place that I found and fixed.

File: library/alloc/src/vec/mod.rs [hunk 4]
```diff
@@ -1023,8 +1023,11 @@ impl<T, A: Allocator> Vec<T, A> {
 
     /// Shrinks the capacity of the vector as much as possible.
     ///
-    /// It will drop down as close as possible to the length but the allocator
-    /// may still inform the vector that there is space for a few more elements.
+    /// The behavior of this method depends on the allocator being used. If the allocator supports
+    /// in-place shrinking, some or all of the excess capacity may be returned to the allocator.
+    /// If not, this method may cause the vector's data to be copied to a new allocation, although
+    /// no guarantees are given that this new allocation is actually smaller. See
+    /// [`Allocator::shrink`] for more details.
```
Comment (the8472) [inline library/alloc/src/vec/mod.rs:1030]: This is correct but occasionally some users misinterpret "no guarantees" and start doing silly things trying to get better behavior.

Instead we can say that the behavior depends on the allocator which may either shrink in-place or reallocate and that the resulting allocation may still have some excess capacity just as `with_capacity` is allowed to.

So it reads more like "performs like copy into Vec::with_capacity(), or better".

Comment (invpt) [reply library/alloc/src/vec/mod.rs]: Thanks, the rephrase does make it sound a little less dire 😄

File: library/alloc/src/vec/mod.rs [hunk 5]
```diff
@@ -3288,10 +3291,12 @@ impl<T, A: Allocator> From<Box<[T], A>> for Vec<T, A> {
 #[cfg(not(test))]
 #[stable(feature = "box_from_vec", since = "1.20.0")]
 impl<T, A: Allocator> From<Vec<T, A>> for Box<[T], A> {
-    /// Convert a vector into a boxed slice.
+    /// Converts a vector into [`Box<[T]>`][owned slice].
```
Comment (the8472) [inline library/alloc/src/vec/mod.rs:3294]: What's the intent behind changing this line?

The [function signature already links](https://doc.rust-lang.org/nightly/std/boxed/struct.Box.html#method.from-23) to `Box` so spelling in plain english seems fine to me and happens to teach the reader an alternate way to call this.

Comment (invpt) [reply library/alloc/src/vec/mod.rs:3294]: Ah, I was just making both doc comments line up with each other since I was already updating part of the From<Vec<...>> docs. I'll revert that specific change.

**Review by the8472 — APPROVED** [review]
Ok looks good now. Can you squash your commits? Then I'll r+ it.
